pkg/scalers/etcd_scaler.go

```diff
@@ -0,0 +1,306 @@
+	}
+
+	meta.etcdKey = config.TriggerMetadata[etcdKey]
+	if len(meta.etcdKey) == 0 {
+		return nil, fmt.Errorf("etcdKey required")
+	}
+
+	meta.value, err = strconv.ParseFloat(config.TriggerMetadata[value], 64)
+	if err != nil || meta.value <= 0 {
+		return nil, fmt.Errorf("value must be a float greater than 0")
+	}
+
+	meta.activationValue = 0
+	if val, ok := config.TriggerMetadata[activationValue]; ok {
+		meta.activationValue, err = strconv.ParseFloat(val, 64)
+		if err != nil {
+			return nil, fmt.Errorf("activationValue must be a float")
+		}
+	}
+
+	meta.watchProgressNotifyInterval = 600
+	if val, ok := config.TriggerMetadata[watchProgressNotifyInterval]; ok {
+		meta.watchProgressNotifyInterval, err = strconv.Atoi(val)
+		if err != nil || meta.watchProgressNotifyInterval <= 0 {
+			return nil, fmt.Errorf("watchProgressNotifyInterval must be a int greater than 0")
+		}
+	}
+
+	if err = parseEtcdAuthParams(config, meta); err != nil {
+		return meta, err
+	}
+
+	meta.scalerIndex = config.ScalerIndex
+	return meta, nil
+}
+
+func getEtcdClients(metadata *etcdMetadata) (*clientv3.Client, error) {
+	var tlsConfig *tls.Config
+	var err error
+	if metadata.enableTLS {
+		tlsConfig, err = kedautil.NewTLSConfigWithPassword(metadata.cert, metadata.key, metadata.keyPassword, metadata.ca)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   metadata.etcdEndpoints,
+		DialTimeout: 5 * time.Second,
+		TLS:         tlsConfig,
+	})
+	fmt.Println(metadata.etcdEndpoints)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to etcd server: %s", err)
+	}
+
+	return cli, nil
+}
+
+// IsActive determines if we need to scale from zero
+func (s *etcdScaler) IsActive(ctx context.Context) (bool, error) {
+	v, err := s.getMetricValue(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	return v > s.metadata.activationValue, nil
+}
+
+// Close closes the etcd client
+func (s *etcdScaler) Close(context.Context) error {
+	return s.client.Close()
+}
+
+// GetMetrics returns value for a supported metric and an error if there is a problem getting the metric
+func (s *etcdScaler) GetMetrics(ctx context.Context, metricName string, metricSelector labels.Selector) ([]external_metrics.ExternalMetricValue, error) {
+	v, err := s.getMetricValue(ctx)
+	if err != nil {
+		return []external_metrics.ExternalMetricValue{}, fmt.Errorf("error getting metric value: %s", err)
+	}
+
+	metric := GenerateMetricInMili(metricName, v)
+	return append([]external_metrics.ExternalMetricValue{}, metric), nil
+}
+
+// GetMetricSpecForScaling returns the metric spec for the HPA.
+func (s *etcdScaler) GetMetricSpecForScaling(context.Context) []v2.MetricSpec {
+	externalMetric := &v2.ExternalMetricSource{
+		Metric: v2.MetricIdentifier{
+			Name: GenerateMetricNameWithIndex(s.metadata.scalerIndex, kedautil.NormalizeString(fmt.Sprintf("etcd-%s-%s", s.metadata.etcdEndpoints, s.metadata.etcdKey))),
```

Comment:
don't use the endpoint here please, the metric name is something internal and using endpoints we could introduce non url characters. IMO we shouldn't use the key either

Reply:
I refer to https://github.com/kedacore/keda/pull/3021#discussion_r873152245 and want to ensure the metric name unique inside the scaler. Could you give me some suggestions for modification?

Reply:
You can use whatever you want because `GenerateMetricNameWithIndex` ensures that the metric is unique inside the ScañedObject, but IDK, maybe somthing like 
```
Name: GenerateMetricNameWithIndex(s.metadata.scalerIndex, kedautil.NormalizeString(fmt.Sprintf("etcd-%s",  s.metadata.etcdKey))),
```
My concern is only about using the endpoints because they could contain any problematic character

Reply:
It seems so. I have just modified it according to your suggestion